File: src/speech_recognition/command_processor.py
#!/usr/bin/env python3
"""
Command processor for Vocalinux.
"""

import logging
import re
from typing import Dict, List, Tuple

# Configure logging
logger = logging.getLogger(__name__)

# Define command patterns
COMMAND_PATTERNS = {
    # Navigation commands
    r"new line|newline|line break": "ACTION:NEW_LINE",
    r"tab|indent": "ACTION:TAB",
    r"(go to|goto) line (\d+)": "ACTION:GOTO_LINE:\\2",
    r"(go|move) (up|down|left|right)( (\d+))?": "ACTION:MOVE_CURSOR:\\2:\\4",
    
    # Editing commands
    r"delete( that)?|remove( that)?": "ACTION:DELETE",
    r"delete (word|line|sentence|paragraph)": "ACTION:DELETE_UNIT:\\1",
    r"delete last (word|line|sentence|paragraph)": "ACTION:DELETE_LAST_UNIT:\\1",
    r"select (word|line|sentence|paragraph)": "ACTION:SELECT_UNIT:\\1",
    r"select all": "ACTION:SELECT_ALL",
    r"copy( that)?": "ACTION:COPY",
    r"cut( that)?": "ACTION:CUT",
    r"paste": "ACTION:PASTE",
    r"undo": "ACTION:UNDO",
    r"redo": "ACTION:REDO",
    
    # Formatting commands
    r"capitalize( that)?": "ACTION:CAPITALIZE",
    r"uppercase|upper case( that)?": "ACTION:UPPERCASE",
    r"lowercase|lower case( that)?": "ACTION:LOWERCASE",
    r"bold( that)?": "ACTION:BOLD",
    r"italic|italics|italicize( that)?": "ACTION:ITALIC",
    r"underline( that)?": "ACTION:UNDERLINE",
    
    # Punctuation commands (these will be directly inserted into text)
    r"^period$|^full stop$": ".",
    r"^comma$": ",",
    r"^question mark$": "?",
    r"^exclamation point$|^exclamation mark$": "!",
    r"^colon$": ":",
    r"^semicolon$": ";",
    r"^quote$|^quotation mark$": "\"",
    r"^single quote$": "'",
    r"^open parenthesis$": "(",
    r"^close parenthesis$": ")",
    r"^open bracket$": "[",
    r"^close bracket$": "]",
    
    # Application control commands
    r"save( file)?|save document": "ACTION:SAVE",
    r"(quit|exit) application": "ACTION:QUIT",
    
    # Vocalinux-specific commands
    r"stop (listening|dictation|recording)": "ACTION:STOP_RECOGNITION",
}


class CommandProcessor:
    """
    Processes speech recognition results for commands.
    
    Detects and processes commands from speech recognition results,
    separating them from regular text input.
    """

    def __init__(self):
        """Initialize the command processor."""
        # Compile all command patterns for efficiency
        self.command_patterns = []
        for pattern, action in COMMAND_PATTERNS.items():
            self.command_patterns.append((re.compile(pattern, re.IGNORECASE), action))
        
    def process_text(self, text: str) -> Tuple[str, List[str]]:
        """
        Process text for commands and regular input.
        
        Args:
            text: The text to process
            
        Returns:
            A tuple of (processed_text, actions)
            - processed_text: Text with commands removed
            - actions: List of action strings to perform
        """
        if not text:
            return ("", [])
            
        # Clean up the text
        text = text.strip()
        
        # Check if the text is entirely a command
        for pattern, action in self.command_patterns:
            match = pattern.match(text)
            if match and match.group(0) == text:
                # The entire text is a command
                if action.startswith("ACTION:"):
                    # This is a special action, not text
                    # Extract any parameters from the match
                    action_parts = action.split(":")
                    action_type = action_parts[1]
                    
                    # Handle parametrized actions
                    if len(action_parts) > 2:
                        # Replace captured groups in the action
                        for i, group in enumerate(match.groups(), 1):
                            if group:
                                action_parts = [
                                    p.replace(f"\\{i}", group) for p in action_parts
                                ]
                        
                        action = ":".join(action_parts)
                    
                    logger.debug(f"Command: {text} -> Action: {action}")
                    return ("", [action])
                else:
                    # This is a replacement text (e.g., punctuation)
                    logger.debug(f"Command: {text} -> Text: {action}")
                    return (action, [])
        
        # Process word by word to find partial commands
        # This is a simpler approach; more sophisticated parsing could be applied
        # for now, just return the full text
        return (text, [])
        
    def get_command_help(self) -> Dict[str, List[str]]:
        """
        Get help information for available commands.
        
        Returns:
            A dictionary of command categories and their commands
        """
        return {
            "Navigation": [
                "new line - Insert a line break",
                "tab - Insert a tab character",
                "go to line [number] - Move cursor to specified line",
                "move [up/down/left/right] [number] - Move cursor in specified direction",
            ],
            "Editing": [
                "delete/remove - Delete selected text or last word/character",
                "delete [word/line/sentence/paragraph] - Delete specified unit",
                "select [word/line/sentence/paragraph] - Select specified unit",
                "select all - Select all text",
                "copy - Copy selected text",
                "cut - Cut selected text",
                "paste - Paste from clipboard",
                "undo - Undo last action",
                "redo - Redo last undone action",
            ],
            "Formatting": [
                "capitalize - Capitalize selected text or next word",
                "uppercase/upper case - Convert to UPPERCASE",
                "lowercase/lower case - Convert to lowercase",
                "bold - Apply bold formatting",
                "italic - Apply italic formatting",
                "underline - Apply underline formatting",
            ],
            "Punctuation": [
                "period/full stop - Insert a period (.)",
                "comma - Insert a comma (,)",
                "question mark - Insert a question mark (?)",
                "exclamation point/mark - Insert an exclamation mark (!)",
                "colon - Insert a colon (:)",
                "semicolon - Insert a semicolon (;)",
                "quote/quotation mark - Insert quotation marks (\"\")",
                "single quote - Insert single quotes ('')",
                "open/close parenthesis - Insert parentheses ()",
                "open/close bracket - Insert brackets []",
            ],
            "Application Control": [
                "save/save file - Save the current file",
                "quit application - Quit the application",
                "stop listening/dictation/recording - Stop voice recognition",
            ],
        }

File: vocalinux-rs/src/speech/command_processor.rs
//! Text command processor for voice commands.

use std::collections::HashMap;

use tracing::debug;

/// Processes recognized text for special commands and formatting.
pub struct CommandProcessor {
    /// Text replacement commands (e.g., "new line" -> "\n")
    text_commands: HashMap<&'static str, &'static str>,
    /// Action commands (e.g., "delete that" -> action)
    action_commands: Vec<&'static str>,
}

impl CommandProcessor {
    pub fn new() -> Self {
        let mut text_commands = HashMap::new();

        // Punctuation
        text_commands.insert("period", ".");
        text_commands.insert("full stop", ".");
        text_commands.insert("comma", ",");
        text_commands.insert("question mark", "?");
        text_commands.insert("exclamation mark", "!");
        text_commands.insert("exclamation point", "!");
        text_commands.insert("colon", ":");
        text_commands.insert("semicolon", ";");
        text_commands.insert("apostrophe", "'");
        text_commands.insert("quote", "\"");
        text_commands.insert("open quote", "\"");
        text_commands.insert("close quote", "\"");
        text_commands.insert("open parenthesis", "(");
        text_commands.insert("close parenthesis", ")");
        text_commands.insert("open bracket", "[");
        text_commands.insert("close bracket", "]");
        text_commands.insert("hyphen", "-");
        text_commands.insert("dash", "-");
        text_commands.insert("underscore", "_");
        text_commands.insert("at sign", "@");
        text_commands.insert("hash", "#");
        text_commands.insert("hashtag", "#");
        text_commands.insert("dollar sign", "$");
        text_commands.insert("percent", "%");
        text_commands.insert("ampersand", "&");
        text_commands.insert("asterisk", "*");
        text_commands.insert("plus sign", "+");
        text_commands.insert("equals sign", "=");
        text_commands.insert("slash", "/");
        text_commands.insert("backslash", "\\");

        // Whitespace and formatting
        text_commands.insert("new line", "\n");
        text_commands.insert("newline", "\n");
        text_commands.insert("new paragraph", "\n\n");
        text_commands.insert("tab", "\t");
        text_commands.insert("space", " ");

        // Action commands that trigger special handling
        let action_commands = vec![
            "delete that",
            "scratch that",
            "undo",
            "undo that",
            "redo",
            "redo that",
            "select all",
            "copy",
            "copy that",
            "cut",
            "cut that",
            "paste",
            "paste that",
            "capitalize",
            "uppercase",
            "lowercase",
        ];

        Self {
            text_commands,
            action_commands,
        }
    }

    /// Process recognized text for commands.
    ///
    /// Returns (processed_text, list_of_actions)
    pub fn process(&self, text: &str) -> (String, Vec<String>) {
        let text_lower = text.to_lowercase();
        let mut actions = Vec::new();

        // Check for action commands first
        for &cmd in &self.action_commands {
            if text_lower.contains(cmd) {
                let action = cmd.replace(' ', "_");
                debug!("Detected action command: {}", action);
                actions.push(action);
            }
        }

        // If an action was found and it's the entire text, return empty text
        if !actions.is_empty() {
            for &cmd in &self.action_commands {
                if text_lower.trim() == cmd {
                    return (String::new(), actions);
                }
            }
        }

        // Process text commands
        let mut result = text.to_string();

        for (&command, &replacement) in &self.text_commands {
            // Case-insensitive replacement
            let pattern = regex_lite::Regex::new(&format!(r"(?i)\b{}\b", regex_lite::escape(command)))
                .unwrap_or_else(|_| regex_lite::Regex::new(command).unwrap());

            result = pattern.replace_all(&result, replacement).to_string();
        }

        // Clean up extra spaces
        let result = result
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ");

        (result, actions)
    }

    /// Get list of available text commands
    pub fn text_commands(&self) -> Vec<&'static str> {
        self.text_commands.keys().copied().collect()
    }

    /// Get list of available action commands
    pub fn action_commands(&self) -> Vec<&'static str> {
        self.action_commands.clone()
    }
}

impl Default for CommandProcessor {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_punctuation() {
        let processor = CommandProcessor::new();
        let (text, actions) = processor.process("Hello period How are you question mark");
        assert_eq!(text, "Hello . How are you ?");
        assert!(actions.is_empty());
    }

    #[test]
    fn test_action_detection() {
        let processor = CommandProcessor::new();
        let (_, actions) = processor.process("delete that");
        assert!(actions.contains(&"delete_that".to_string()));
    }

    #[test]
    fn test_new_line() {
        let processor = CommandProcessor::new();
        let (text, _) = processor.process("First line new line Second line");
        assert!(text.contains('\n'));
    }
}
